lustrefs-exporter/src/lib.rs

```diff
@@ -62,26 +113,384 @@ impl LabelProm for TargetVariant {
+impl<T> ShouldRegister for Family<T> {
+    fn should_register(&self) -> bool {
+        self.is_set
+    }
+}
+
+/// Trait for registering metrics with a Prometheus registry.
+///
+/// This trait provides a standardized interface for metric collections to register
+/// themselves with a Prometheus registry. Implementors should register all their
+/// contained metrics using the appropriate registration methods.
+///
+/// The trait is typically implemented by metric collection structs that contain
+/// multiple related metrics (e.g., `HostMetrics`, `LNetMetrics`, `StatsMetrics`).
+/// Each implementation should use conditional registration via `maybe_register`
+/// or `maybe_register_without_auto_suffix` to only register metrics that contain data.
+///
+/// # Examples
+///
+/// ```
+/// use lustrefs_exporter::{Family, RegisterMetrics, MaybeRegister as _};
+/// use prometheus_client::{metrics::counter::Counter, registry::Registry};
+///
+/// #[derive(Default)]
+/// struct MyMetrics {
+///     requests_total: Family<Counter<u64>>,
+/// }
+///
+/// impl RegisterMetrics for MyMetrics {
+///     fn register_metric(&self, registry: &mut Registry) {
+///         registry.maybe_register(
+///             "requests_total",
+///             "Total number of requests",
+///             &self.requests_total,
+///         );
+///     }
+/// }
+/// ```
+pub trait RegisterMetrics {
+    fn register_metric(&self, registry: &mut Registry);
+}
+
+/// Trait for creating or retrieving metrics from a metric family.
+///
+/// This trait provides a standardized interface for getting or creating metrics within
+/// a metric family based on a set of labels. If a metric with the given labels already
+/// exists, it returns a reference to that metric. If not, it creates a new metric with
+/// those labels and returns a reference to it.
+///
+/// It's important to note that this trait is designed to work with the Famil<T> wrapper
+/// and will automatically mark itself as being set so that it will be registered with the
+/// Prometheus registry. It returns the response from the inner `get_or_create` call,
+/// allowing the user to continue working with the internal metric. This allows the code to
+/// look the same as if a wrapper was not being used, maintaining code uniformity.
+///
+/// # Type Parameters
+///
+/// * `T` - The type of metric being stored in the family (e.g., `Counter<u64>`, `Gauge<f64>`)
+///
+/// # Examples
+///
+/// ```
+/// use lustrefs_exporter::{Family, GetOrCreate as _, create_labels};
+/// use prometheus_client::metrics::counter::Counter;
+///
+/// let mut family: Family<Counter<u64>> = Family::default();
+///
+/// let labels = create_labels(&[("target", "fs-OST0000".to_string())]);
+/// let counter = family.get_or_create(&labels).inc();
+/// ```
+pub trait GetOrCreate<T> {
+    fn get_or_create(
+        &mut self,
+        labels: &LabelContainer,
+    ) -> lock_api::MappedRwLockReadGuard<'_, parking_lot::RawRwLock, T>;
+}
```

Comment:
Why is this a trait ? The original is a method of the struct. Why not implement it generically ?